### PR TITLE
TypeSchemes for OpDefs (by adding {TypeArg,PrimType}::Variable)

### PR DESCRIPTION
**Pull request by acl-cqc**
Ok here goes. The limited form where polymorphism is limited to a special `OpDefTypeScheme` that is not used in core Hugr, but provides a `compute_signature` function that can be used by an `OpDef`. (But, where we add type variables to the core `Type` and `TypeArg` hierarchies to enable this.) I've now replaced the template `(String, String)` SignatureFunc with this schema (parsing can go somewhere else), and included #541 in here (will rebase/merge on top when/if that goes in)

See what you think as it stands - I maintain that I don't think this PR is a good endpoint, but I'll try to do the rest in a separate follow-up PR.

That is, I think
* If we are adding new `Variable`s here, we should go the whole way and make FunctionType polymorphic. This is not difficult given what's already in this PR:
    1. we move the `Vec<TypeParam>` from OpDefTypeScheme into FunctionType (so OpDefTypeScheme is just a FunctionType) - or have OpDefTypeScheme as a separate PrimType enum-variant;
    2. whereas this PR has OpDefTypeScheme::new doing validation using only the type params in the OpDefTypeScheme, and FunctionType::validate using only type params provided from outside; instead the new (Function/)Type takes a bunch of type parameters from outside *and* concatenates them with the ones on the inside (for the recursive calls only)
    3. `Call` node takes (as constructor) inputs outputs extension (and no type params) - this is the argument for keeping `FunctionType` and the polymorphic variant separate in the Rust types.
    4. Add an optype that does type application - constructed with a polymorphic (Function/)Type and a list of TypeArgs, and the constructor checks the type-args against the type-params declared in the FunctionType. (This is the bit of "dependent types" that we can't do if we implement this in an Extension).
* Alternatively, if we really don't want polymorphic FunctionTypes, I think it's plausible to use Opaques rather than adding new enum-elements for Variable, with custom callbacks for handling Opaques rather than lists of TypeParams/TypeArgs. This'd be a lot more generic in the core, and keep type-scheme-specific code in the type-scheme module.

### CODE REVIEW

File: src/extension.rs [hunk 1]
```diff
@@ -324,31 +324,39 @@ pub enum ExtensionBuildError {
...
+    }
+
     /// Returns `true` if the set contains the given extension.
     pub fn contains(&self, extension: &ExtensionId) -> bool {
         self.0.contains(extension)
     }
 
     /// Returns `true` if the set is a subset of `other`.
     pub fn is_subset(&self, other: &Self) -> bool {
+        if !other.1.is_empty() {
```
Comment (ss2165) [inline src/extension.rs:357]: the number of these panics suggests there are two different impls here squeezed in to one.

**Review by ss2165 — COMMENTED** [review]
looks pretty nice broadly.

I think _if_ we go with polymorphic functions, I prefer `FunctionType` becoming polymorphic, and "concrete" function type being a dynamic concept i.e. (a dynamic query to the function type to ask if there are any free variables).

File: src/types.rs [hunk 2]
```diff
@@ -299,6 +299,35 @@ impl Type {
             }
         }
     }
+
+    /// Could make this public, but easier to panic if the TypeArg's don't match the TypeParams
+    /// (should be checked when the args are first given to the type scheme).
+    pub(crate) fn substitute(&self, args: &[TypeArg]) -> Self {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types.rs]: I tried this, but found it barely helped (if at all), and has to be less efficient - clone first clones *everything*, then builds some new bits where the old bits have to change; substitute-takes-ref only clones the bits that are common to the original, and then builds the same "new bits". See dd00c63220d85449773141a0320ff1a1bc7f4e58 if you want to take a look (but I've then reverted).

Moreover, the common case is in a type scheme's `compute_signature`, in which case we can't consume the type scheme, so we'd have to clone. (Possibly at some point we may use this for substituting across a whole Hugr, not just the types, and *possibly* the code to do that could be written as a destructive mutator rather than a copy-and-create, that might fit well with a Tierkreis/rust-style implementation, so there might be some benefit there in being able to consume the input Hugr. Seems very speculative at this point though.)

File: src/types.rs [hunk 3]
```diff
@@ -246,6 +247,13 @@ impl Type {
         Self(TypeEnum::Sum(SumType::Simple { size }), TypeBound::Eq)
     }
 
+    /// New type variable (for use in type schemes only),
+    /// with bound matching that in the type scheme
+    /// (i.e. the variable must be declared as a [TypeParam::Type])
+    pub fn new_variable(idx: usize, bound: TypeBound) -> Self {
```
Comment (acl-cqc) [inline src/types.rs:253]: Hmmm, this should possibly be called `use_var` (a) for consistency with TypeArg::use_var, and (b) to emphasize that it doesn't *create* (i.e. declare) a new variable, merely a new reference/usage of an existing variable (declared elsewhere). `new_variable` may be consistent with `Type::new_extension`, `Type::new_function` etc., but is potentially misleading.

File: src/extension.rs [hunk 4]
```diff
@@ -343,6 +342,14 @@ impl ExtensionSet {
         self.0.insert(extension.clone());
     }
 
+    /// Adds a type var (which must have been declared as a [TypeParam::Extensions]) to this set
+    pub fn insert_type_var(&mut self, idx: usize) {
+        // Represent type vars as string representation of DeBruijn index.
+        // This is not a legal IdentList or ExtensionId so should not conflict.
```
Comment (acl-cqc) [inline src/extension.rs:348]: Yes, this is a hack...I tried having two HashSet's in 80d8d8d4dad110f8b7d5afea9a7e449aeff0429c but it was messier (and caused a mass of clippy errors because of making error types (that contain ExtensionSets) too big - so we'd need to add some Boxes too)

File: src/types/signature.rs [hunk 5]
```diff
@@ -59,6 +60,26 @@ impl FunctionType {
     pub fn pure(self) -> Signature {
         self.with_input_extensions(ExtensionSet::new())
     }
+
+    pub(crate) fn validate(
```
Comment (acl-cqc) [inline src/types/signature.rs:64]: This was just moved out of `types.rs` in e5ec9d64393966930966e0cf7ced6fb03dd10858, the only "new" bit is adding the `type_vars` parameter

File: src/types/type_param.rs [hunk 6]
```diff
@@ -84,35 +102,110 @@ pub enum TypeArg {
         arg: CustomTypeArg,
     },
     /// Instance of [TypeParam::List] or [TypeParam::Tuple], defined by a
-    /// sequence of arguments.
+    /// sequence of elements.
     Sequence {
         #[allow(missing_docs)]
-        args: Vec<TypeArg>,
+        elems: Vec<TypeArg>,
```
Comment (acl-cqc) [inline src/types/type_param.rs:108]: This was a driveby, see separate commit 405d362e1ba63de3009a6134a2141d65580bc054

File: src/types/type_param.rs [hunk 7]
```diff
@@ -84,35 +102,110 @@ pub enum TypeArg {
...
     },
     /// Instance of [TypeParam::Extensions], providing the extension ids.
     Extensions {
         #[allow(missing_docs)]
         es: ExtensionSet,
     },
+    /// Variable (used in type schemes only)
+    Variable {
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply src/types/type_param.rs:116]: Yes, this would make it more difficult to capture the serialised version with a pydantic model

File: src/extension.rs [hunk 8]
```diff
@@ -324,31 +324,39 @@ pub enum ExtensionBuildError {
 
 /// A set of extensions identified by their unique [`ExtensionId`].
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension.rs]: Gone with malformed `ExtensionId`s instead...hackier but less intrusive

File: src/extension.rs (hunk 1 above)
Comment (acl-cqc) [reply src/extension.rs]: Gone with malformed ExtensionIds instead...hackier but less intrusive

File: src/extension/type_scheme.rs [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    fn id_fn(t: Type) -> FunctionType {
+        FunctionType::new(vec![t.clone()], vec![t])
+    }
+
+    #[test]
+    fn test_mismatched_args() -> Result<(), SignatureError> {
+        let ar_def = PRELUDE.get_type("array").unwrap();
+        let typarams = [TypeParam::Type(TypeBound::Any), TypeParam::max_nat()];
+        let tyvar = TypeArg::use_var(0, typarams[0].clone());
+        let szvar = TypeArg::use_var(1, typarams[1].clone());
```
Comment (croyzor) [inline src/extension/type_scheme.rs:108]: Is this what they call Polish Notation? 🤪

File: src/extension/type_scheme.rs [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+            ))
+        );
+
+        // (Try to) make a schema with bad args
+        let arg_err = SignatureError::TypeArgMismatch(TypeArgError::TypeMismatch {
+            param: typarams[0].clone(),
+            arg: szvar.clone(),
+        });
+        assert_eq!(
+            ar_def.instantiate_concrete([szvar.clone(), tyvar.clone()]),
```
Comment (croyzor) [inline src/extension/type_scheme.rs:141]: I find it a bit disturbing that the collections extension seems to have its own distinct notion of type variables

File: src/extension/type_scheme.rs [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    fn test_misused_variables() -> Result<(), SignatureError> {
+        // Variables in args have different bounds from variable declaration
+        let tv = TypeArg::use_var(0, TypeParam::Type(TypeBound::Copyable));
+        let list_def = EXTENSION.get_type(&LIST_TYPENAME).unwrap();
+        let body_type = id_fn(Type::new_extension(list_def.instantiate_concrete([tv])?));
+        let reg = [EXTENSION.to_owned()].into();
+        for decl in [
+            TypeParam::Extensions,
+            TypeParam::List(Box::new(TypeParam::max_nat())),
+            TypeParam::Type(TypeBound::Any),
```
Comment (croyzor) [inline src/extension/type_scheme.rs:167]: Might as well include the other constructors of `TypeParam`

File: src/types.rs [hunk 12]
```diff
@@ -271,23 +279,79 @@ impl Type {
     pub(crate) fn validate(
```
Comment (croyzor) [inline src/types.rs:279]: This function could use a docstring

File: src/types.rs [hunk 13]
```diff
@@ -271,23 +279,79 @@ impl Type {
...
+            TypeEnum::Prim(PrimType::Extension(custy)) => {
+                custy.validate(extension_registry, type_vars)
+            }
+            TypeEnum::Prim(PrimType::Function(ft)) => ft.validate(extension_registry, type_vars),
+            TypeEnum::Prim(PrimType::Variable(idx, bound)) => {
+                if type_vars.get(*idx) == Some(&TypeParam::Type(*bound)) {
+                    Ok(())
+                } else {
+                    Err(SignatureError::TypeVarDoesNotMatchDeclaration {
+                        decl: type_vars.get(*idx).cloned(),
```
Comment (croyzor) [inline src/types.rs:301]: This message isn't so helpful in the case where there's a `Var(n + 1)` and `n` type args (i.e. `type_vars.get(...)` is `None`)

File: src/extension/type_scheme.rs [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+        Ok(Self { params, body })
+    }
+
+    pub(super) fn compute_signature(
+        &self,
+        args: &[TypeArg],
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<FunctionType, SignatureError> {
+        check_type_args(args, &self.params)?;
+        Ok(self.body.substitute(extension_registry, args))
```
Comment (croyzor) [inline src/extension/type_scheme.rs:49]: We need to validate that the `args` themselves are closed before we call `substitute`

**Review by croyzor — COMMENTED** [review]
Looks good! My main issue is the lack of validation that type args which are being substituted don't have any free vars (which I think is a property we want). Personally, I'd like to add an extra struct like `ClosedTypeArg` so that the typechecker can enforce this difference.
(Say, the only way to construct a `ClosedTypeArg` is via `TypeArg.check_closed()` or a substitution).

I'm also not really sure what the relationship is between the substitution here and the "instantiation" that happens in the collections extension, which I think are separate implementations of the same thing

File: src/hugr/validate.rs [hunk 15]
```diff
@@ -164,7 +164,8 @@ impl<'a, 'b> ValidationContext<'a, 'b> {
         //   - see https://github.com/CQCL-DEV/hugr/issues/508 )
         if let OpType::LeafOp(crate::ops::LeafOp::CustomOp(b)) = op_type {
             for arg in b.args() {
-                arg.validate(self.extension_registry)
+                // Hugrs are monomorphic, so no type variables in scope
+                arg.validate(self.extension_registry, &[])
```
Comment (acl-cqc) [inline src/hugr/validate.rs:168]: @croyzor see here for where we validate TypeArgs are closed

File: src/extension.rs [hunk 16]
```diff
@@ -381,6 +396,35 @@ impl ExtensionSet {
     pub fn iter(&self) -> impl Iterator<Item = &ExtensionId> {
         self.0.iter()
     }
+
+    pub(crate) fn validate(&self, params: &[TypeParam]) -> Result<(), SignatureError> {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension.rs:401]: Ooh, that's an interesting idea. I do rather like that...

OTOH we have also said that we want to be able to deal with code where we *don't* have the extensions available. For example, we have [`ExternalOp::Opaque`](https://github.com/CQCL-DEV/hugr/blob/ebb2473385f426212547e5a2bb059cd69967fae5/src/ops/custom.rs#L27).

But if we can pass extension *declarations* around (even when we can't run the ops, etc.) then maybe we can make this requirement, hmmm. @ss2165 any guidance here?

File: src/extension/type_scheme.rs [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+    use crate::types::{CustomType, FunctionType, Type, TypeBound};
+    use crate::Extension;
+
+    use super::OpDefTypeScheme;
+
+    #[test]
+    fn test_opaque() -> Result<(), SignatureError> {
+        let list_def = EXTENSION.get_type(&LIST_TYPENAME).unwrap();
+        let tyvar = TypeArg::use_var(0, TypeParam::Type(TypeBound::Any));
+        let list_of_var = Type::new_extension(list_def.instantiate_concrete([tyvar.clone()])?);
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension/type_scheme.rs:78]: I agree ;-). However, `instantiate_concrete` is the right choice of name for OpDefs, for now (while Hugrs are monomorphic), even if not TypeDefs (where "instantiate" would be better - the point is that even if you instantiate with a variable, you are converting from a polymorphic thing `forall a.....a....` to a monomorphic thing with no `forall`.)

Giving these different names would mean ending the `TypeParametrised` trait, but then if we go for polymorphic Hugrs, we'd probably want to bring that trait back (but rename the method `instantiate` as then both OpDefs and TypeDefs would be able to be instantiated with type variables that are free/from an outer scope). So I'm not sure about the churn, but I'll look into @ss2165's `TypeParametrised` a bit more and have a think.

File: src/extension/type_scheme.rs (hunk 9 above)
Comment (acl-cqc) [reply src/extension/type_scheme.rs]: Yes ;). Just for that, I've combined the redundant code in the two `let`s ;-)

File: src/extension/type_scheme.rs (hunk 10 above)
Comment (acl-cqc) [reply src/extension/type_scheme.rs:143]: Hmmm, I'm not sure what you mean here?

Quite a few extensions use custom binary functions to do things that this type-scheme system will do for them, so a follow-up PR is definitely to remove all that binary code and use type schemes. However, I don't think that needs to be part of this PR.

File: src/extension/type_scheme.rs (hunk 11 above)
Comment (acl-cqc) [reply src/extension/type_scheme.rs]: Yes, fair, done. But I've removed some that are redundant given the more-detailed checks in `test_bound_covariance`

File: src/types.rs (hunk 12 above)
Comment (acl-cqc) [reply src/types.rs:286]: Yes, good point :). I've done `TypeArg::validate` as well as that's also `pub(crate)` but the other `fn validate`s are less visible so I haven't done so there.

File: src/extension/type_scheme.rs (hunk 14 above)
Comment (acl-cqc) [reply src/extension/type_scheme.rs:52]: Yeah ok, done - with a comment, 'coz this'll go when/if we make Hugr's polymorphic, but the check in `hugr/validate.rs` should be able to stay (using the Hugr's bound variables when it has them)

File: src/extension/type_scheme.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension/type_scheme.rs]: Sorry, I tell a lie. I see now that `instantiate_concrete` is not part of `TypeParametrised` at all; it is defined for TypeDef only, whereas the OpDef "equivalent" (no common code) is called `instantiate_opaque`. So yes, renaming `instantiate_concrete` to `instantiate` is all good, done! :-)

File: src/extension.rs [hunk 18]
```diff
@@ -381,6 +396,35 @@ impl ExtensionSet {
...
+            Some(i) => match args.get(i) {
+                Some(TypeArg::Extensions{es}) => es.iter().cloned().collect::<Vec<_>>(),
+                _ => panic!("value for type var was not extension set - type scheme should be validate()d first"),
+            },
+        }))
+    }
+}
+
+fn as_typevar(e: &ExtensionId) -> Option<usize> {
+    match e.chars().next() {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension.rs:422]: I guess I could even check whether the Id was legal and rely on it parsing as an int if not, but that'd be a lot more expensive (regex etc.)

File: src/types/type_param.rs [hunk 19]
```diff
@@ -84,35 +102,110 @@ pub enum TypeArg {
...
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+pub struct TypeArgVariable {
+    idx: usize,
+    cached_decl: TypeParam,
 }
 
 impl TypeArg {
+    /// Makes a TypeArg representing the type variable with the specified (DeBruijn) index
+    /// and declared [TypeParam].
+    pub fn use_var(idx: usize, decl: TypeParam) -> Self {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types/type_param.rs]: Renamed both `use_var` and `new_variable` to `new_var_use`, hope that makes it clear

File: src/types.rs (hunk 3 above)
Comment (acl-cqc) [reply src/types.rs]: I've gone with `new_var_use` in both, hope that makes it clear

File: src/extension/type_scheme.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension/type_scheme.rs]: Possibly this should be `instantiate_type` or `..._custom` or `..._external` or something - in #542 we'll also have a method on polymorphic function types that will want another distinct name too...

File: src/extension.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (ss2165) [reply src/extension.rs:401]: If we can pass declarations around then we don't need the opaque op case where the signature is serialized right? We decided we did need that (not sure I'm remembering exactly why)?

[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension.rs:401]: Well, the declaration would not (could not!) include the binary `compute_signature`, hmmm. So some kind kind of partial declaration...

[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension.rs:401]: (We could say that it includes the types, at least. But I'm not sure whether we'd want to go that route. @croyzor?)

[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension.rs:401]: Hmmmm, hang on. We have been assuming that (when we validate a type, for example) that we do have at least the declarations of the type in each extension available. This does not mean we have op defs, let alone binary `compute_signature`s, but maybe we need to put this on a firmer footing. (In another PR I suspect)

File: src/types/type_param.rs [hunk 20]
```diff
@@ -84,35 +103,103 @@ pub enum TypeArg {
...
     },
     /// Instance of [TypeParam::Extensions], providing the extension ids.
     Extensions {
         #[allow(missing_docs)]
         es: ExtensionSet,
     },
+    /// Variable (used in type schemes only) - see [TypeArg::new_var_use]
+    Variable {
+        #[allow(missing_docs)]
+        v: TypeArgVariable,
```
Comment (mark-koch) [inline src/types/type_param.rs:119]: Is there a reason why you declare the `TypeArgVariable` struct instead of just having `idx` and `chached_decl` here?

**Review by mark-koch — COMMENTED** [review]
Nice work, looks good to me! And sorry for being so late with this review

As a side note: While thinking about generic types I was wondering if we're interested in also having contravariant and invariant generic types at some point?

File: src/types.rs [hunk 21]
```diff
@@ -268,24 +276,103 @@ impl Type {
...
+    /// of bound variables, and that for each [CustomType] the corresponding
+    /// [TypeDef] is in the [ExtensionRegistry] and the type arguments
+    /// [validate] and fit into the def's declared parameters.
+    ///
+    /// [validate]: crate::types::type_param::TypeArg::validate
+    /// [TypeDef]: crate::extension::TypeDef
     pub(crate) fn validate(
         &self,
         extension_registry: &ExtensionRegistry,
+        type_vars: &[TypeParam],
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types.rs:289]: Good point. It's meant to be the declarations, i.e. an explanation of what each variable is, rather than values of the variables. I'll rename to `var_decls` (? - could be `declared_variables` or `bound_vars` or `vars_in_scope` ?)

File: src/types.rs [hunk 22]
```diff
@@ -256,6 +257,13 @@ impl Type {
         Self(TypeEnum::Sum(SumType::Simple { size }), TypeBound::Eq)
     }
 
+    /// New type variable (for use in type schemes only),
+    /// with bound matching that in the type scheme
+    /// (i.e. the variable must be declared as a [TypeParam::Type])
+    pub fn new_var_use(idx: usize, bound: TypeBound) -> Self {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types.rs:263]: Yeah, so I want to make clear that it's not creating a new variable (i.e. declaring), where `new_var` seems a bit ambiguous. (It's only `new` in the Rust constructing sense.) Maybe `ref` or `occurrence`? I found this a bit awkward, naming ideas most welcome

File: src/types/type_param.rs (hunk 20 above)
Comment (acl-cqc) [reply src/types/type_param.rs:119]: One of Rust's annoyances is that to match on the enum you have to make it `pub`, which makes the whole structure (all variants and their fields) public - I mean, it has to, really - and then anyone can create instances like `TypeArg::Variable(2, TypeParam::Type(TypeBound::Eq))`. This is a malformed instance, as it needs to be represented as `TypeArg::Type(Type::new_variable(2, TypeBound::Eq))`.

I think you'd figured that out already. Of course we could detect malformed instances only in `validate` and that would still be OK - this just prevents them from ever being constructed i.e. detects them much eariler.

File: src/types.rs [hunk 23]
```diff
@@ -268,24 +276,103 @@ impl Type {
...
+    idx: usize,
+    used_as: &TypeParam,
+) -> Result<(), SignatureError> {
+    match decls.get(idx) {
+        None => Err(SignatureError::FreeTypeVar {
+            idx,
+            num_decls: decls.len(),
+        }),
+        Some(decl) => {
+            if decl == used_as {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types.rs:378]: Yeah this is a bit tricky. The field inside `TypeArgVariable` is intended as a strict cache of the declaration, and then we check it fits (with contains) in `check_type_arg`. So, the `cached_decl` is the widest thing that fits within the decl. Alternatively, we could require equality in `check_type_arg`, and have a containment check here; again, that would mean there was exactly one correct value for the cached_decl, just the opposite way around (it would be the narrowest thing that contains the usage, essentially).

Or we could even allow containment at both ends, and then `cached_decl` could be *anything* that fits in the decl and contains the usage - there could be multiple "legal" values at each usage site. I dislike the latter just because it's less precise...

The *ideal* would be to avoid `cached_decl` altogether - which would be possible, but every API that takes `TypeArg`s would have to have an extra parameter (the `TypeParam` decl's of any variable that might occur in those `TypeArg`s). Specifically, that includes `check_type_arg`(`s`) and friends, which already take a `TypeParam` (the thing to check whether the TypeArg fits into), so then check_type_arg(s) would take two TypeParam-things, which invites confusion as to which is which....and `check_type_arg`(`s`) is used in quite a lot of places, so sadly I think I have to dismiss that one :-(.

[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types.rs:378]: I've done a bit of renaming, hopefully that's clearer. I can add a comment here (that the containment check is done in `check_type_arg`) if you think that'd help?

File: src/types/type_param.rs [hunk 24]
```diff
@@ -84,35 +103,103 @@ pub enum TypeArg {
...
         #[allow(missing_docs)]
-        args: Vec<TypeArg>,
+        elems: Vec<TypeArg>,
     },
     /// Instance of [TypeParam::Extensions], providing the extension ids.
     Extensions {
         #[allow(missing_docs)]
         es: ExtensionSet,
     },
+    /// Variable (used in type schemes only) - see [TypeArg::new_var_use]
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types/type_param.rs:116]: Right, wasn't sure whether the comment on `TypeArgVariable` would be enough, done :)

File: src/types/type_param.rs [hunk 25]
```diff
@@ -61,6 +62,24 @@ impl TypeParam {
...
     }
+
+    fn contains(&self, other: &TypeParam) -> bool {
+        match (self, other) {
+            (TypeParam::Type(b1), TypeParam::Type(b2)) => b1.contains(*b2),
+            (TypeParam::BoundedNat(b1), TypeParam::BoundedNat(b2)) => match (b1.0, b2.0) {
+                (None, _) => true,
+                (Some(b1), Some(b2)) if b1 >= b2 => true,
+                _ => false,
+            },
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types/type_param.rs]: Good call, done

File: src/extension/op_def.rs [hunk 26]
```diff
@@ -91,25 +92,23 @@ pub trait CustomLowerFunc: Send + Sync {
...
 #[derive(serde::Deserialize, serde::Serialize)]
 pub(super) enum SignatureFunc {
-    // Note: I'd prefer to make the YAML version just implement the same CustomSignatureFunc trait,
-    // and then just have a Box<dyn CustomSignatureFunc> instead of this enum, but that seems less likely
-    // to serialize well.
-    /// TODO: these types need to be whatever representation we want of a type scheme encoded in the YAML
+    // Note: except for serialization, we could have type schemes just implement the same
+    // CustomSignatureFunc trait too, and replace this enum with Box<dyn CustomSignatureFunc>.
+    // However instead we treat all CustomFunc's as non-serializable.
     #[serde(rename = "signature")]
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/extension/op_def.rs:98]: Yeah, I was copying the YAML fields in the spec, but perhaps that is looking too far ahead ;). If I change the spec to say `type_scheme`, can I drop the serde-rename here altogether? Or would the YAML then have to use CamelCase?

File: src/types.rs [hunk 27]
```diff
@@ -268,24 +276,103 @@ impl Type {
...
+    /// contains a type with an incorrect [TypeBound], or there are not enough `args`.
+    /// These conditions can be detected ahead of time by [Type::validate]ing against the [TypeParam]s
+    /// and [check_type_args]ing the [TypeArg]s against the [TypeParam]s.
+    pub(crate) fn substitute(&self, exts: &ExtensionRegistry, args: &[TypeArg]) -> Self {
+        match &self.0 {
+            TypeEnum::Prim(PrimType::Alias(_)) | TypeEnum::Sum(SumType::Simple { .. }) => {
+                self.clone()
+            }
+            TypeEnum::Prim(PrimType::Variable(idx, bound)) => match args.get(*idx) {
+                Some(TypeArg::Type { ty }) => ty.clone(),
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply src/types.rs:341]: I see, I guess you could put an assert here but not strictly necessary. I was just wondering since you caught the other errors explicitly

File: src/types.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (mark-koch) [reply src/types.rs]: I like `var_decls` 👍

File: src/types.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (mark-koch) [reply src/types.rs]: I see, the renaming makes things more clear 👍 I think adding a comment pointing to `check_type_arg` wouldn't be a bad idea

File: src/types.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (mark-koch) [reply src/types.rs:267]: Ahh, I was thinking `new` in the Rust constructing sense, similar to `new_function`, `new_tuple` etc. In that context I wouldn't find `new_var` too ambiguous, but I'm also fine with keeping `new_var_use`

**Review by mark-koch — APPROVED** [review]

File: src/types.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types.rs]: Ok, done

File: src/types.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types.rs:267]: I've improved the docs. I'm a bit tempted to rename to `new_var_ref` but that could be confusing too so leaving as is.

**Comment by acl-cqc** [comment]
Going straight for #630 ...
